### PR TITLE
[#1066] Pass all participating projects along with BuildData

### PR DESCRIPTION
**Pull request by szarnekow**
[#1066] Pass all participating projects along with BuildData

### CODE REVIEW

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuildData.java [hunk 1]
```diff
@@ -45,7 +57,26 @@ public BuildData(String projectName, ResourceSet resourceSet, ToBeBuilt toBeBuil
...
+		this(participatingProjectNames.iterator().next(), resourceSet, toBeBuilt, queuedBuildData, indexingOnly, buildRequestor, participatingProjectNames);
+	}
+	
+	/**
+	 * @since 2.18
+	 */
+	public BuildData(String projectName, ResourceSet resourceSet, ToBeBuilt toBeBuilt, QueuedBuildData queuedBuildData, boolean indexingOnly, Runnable buildRequestor, Set<String> participatingProjectNames) {
 		this.projectName = projectName;
+		if (participatingProjectNames.contains(projectName) || projectName == null) {
+			this.participatingProjectNames = ImmutableSet.copyOf(participatingProjectNames);
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuildData.java]: am not sure if i like the mix of guava and std java api for set immuntability

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 2]
```diff
@@ -319,15 +320,62 @@ protected void incrementalBuild(IResourceDelta delta, final IProgressMonitor mon
...
 
-	private void addInfosFromTask(Task task, final ToBeBuilt toBeBuilt) {
+	private static Boolean mustCallDeprecatedDoBuild = null;
+	private static boolean wasDeprecationWarningLogged = false;
+	
+	/**
+	 * @since 2.18
+	 */
+	protected void addInfosFromTaskAndBuild(Task task, final ToBeBuilt toBeBuilt, BuildType buildType, IProgressMonitor monitor)
+			throws CoreException {
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: highly sophisticated :)

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 3]
```diff
@@ -319,15 +320,62 @@ protected void incrementalBuild(IResourceDelta delta, final IProgressMonitor mon
...
+				doBuild(toBeBuilt, task.getProjectNames(), monitor, buildType);
+			}
+		}
+		
+	}
+
+	/**
+	 * @since 2.18
+	 */
+	protected boolean isDoBuildSpecialized(Class<?> c) {
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: i wonder if we can do something not to keep this stuff forever e.g. announcing a removal in 2020

Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: I'll mark it as deprecated immediately.

Comment (cdietrich) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: (same for the others that are for compatibility)

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java (hunk 2 above)
Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: I tried to break only few things.

What I was also thinking about: changing the constructor of BuildData to smth like `BuildData(Map<Key<T>, T>)` to become more forward compatible.

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java (hunk 3 above)
Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: I think I caught the relevant candidates.

**Comment by szarnekow** [comment]
@cdietrich What do you think?

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/ProjectOpenedOrClosedListener.java [hunk 4]
```diff
@@ -263,15 +263,22 @@ protected void processClosedProjects(IProgressMonitor monitor) {
...
 		String taskName = Messages.ProjectOpenedOrClosedListener_RemovingProject + projectNames
 				+ Messages.ProjectOpenedOrClosedListener_FromIndex;
 		monitor.setTaskName(taskName);
 		SubMonitor progress = SubMonitor.convert(monitor, 1);
 		try {
 			ResourceSet resourceSet = createResourceSet();
-			BuildData buildData = new BuildData(null, resourceSet, task.getToBeBuilt(), queuedBuildData, false,
-					BuildManagerAccess::requestBuild);
+			BuildData buildData = new BuildData(
+					"",
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/ProjectOpenedOrClosedListener.java]: null?

Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/ProjectOpenedOrClosedListener.java]: If you use `workspace.getProject(buildData.getProjectName())` you'd get an assertion error. With the empty string, you'll find a project that does not exist. We do have a test that fails if we use null here. I'm not really passionate - changing it into one or the other direction would be ok'ish. But the empty string is more failsafe I suppose.

Comment (cdietrich) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/ProjectOpenedOrClosedListener.java]: ok. can you please add a javadoc comment to BuildData.getProjectName

**Pull request merged**